### ISSUE TITLE
[configuration] `disabled` boolean flags are inconsistent with other parts of the OpenTelemetry ecosystem

### ISSUE THREAD

**Issue by mx-psi** [open-telemetry/opentelemetry-specification#4344]
edit (2025-11-05): I am making a summary of the positions so far as well as the arguments in favor or against. The text on the issue before this date is left under the line. Verbatim quotes are formatted in italics.

### Description

This is what it is currently done for environment variables (see [v1.50.0 version of the spec](https://github.com/open-telemetry/opentelemetry-specification/blob/v1.50.0/specification/configuration/sdk-environment-variables.md#boolean)):

>  All Boolean environment variables SHOULD be named and defined such that false is the expected safe default behavior. Renaming or changing the default value MUST NOT happen without a major version upgrade.

The declarative configuration spec follows this guidance.

### Usages

This convention is used in 9 places in the declarative configuration schema v1.0.0-rc.1 (see [kitchen sink](https://github.com/open-telemetry/opentelemetry-configuration/blob/v1.0.0-rc.2/examples/kitchen-sink.yaml)).

Some of the usages in the Collector, Java Agent and .NET instrumentation accidentally follow this guidance insofar as they are enabled-by-default. See https://github.com/open-telemetry/opentelemetry-specification/issues/4344#issuecomment-3416330281, https://github.com/open-telemetry/opentelemetry-specification/issues/4344#issuecomment-3421326691 and https://github.com/open-telemetry/opentelemetry-specification/issues/4344#issuecomment-3460473132 for full details.

### Description

All boolean configuration flags should be named `enabled` regardless of the default value.

### Usages

The Collector, Java Agent and .NET instrumentation de facto follow this position. See https://github.com/open-telemetry/opentelemetry-specification/issues/4344#issuecomment-3416330281, https://github.com/open-telemetry/opentelemetry-specification/issues/4344#issuecomment-3421326691 and https://github.com/open-telemetry/opentelemetry-specification/issues/4344#issuecomment-3460473132 for full details. There is one exception for the Collector (but, I argue, it is an 'accident' and not intentional).

### Description

There should be a separate option (e.g. `min_stability`) that changes the defaults of instrumentations and other features depending on what stability level we consider them having.

### Usages

Not currently used.

## Arguments in favor and against different positions

Arguments **in favor of Position A in the context of environment variables** can be found mostly on https://github.com/open-telemetry/opentelemetry-specification/pull/2755#discussion_r962171923. As a summary the arguments espoused were the following (see full comment and context [here](https://github.com/open-telemetry/opentelemetry-specification/pull/2755#discussion_r970047515)):

- *"User confusion around if they are or are not required to provide an environment variable diminishes when they know that by not providing it it will be assumed false"*
- *"It would be a huge frustration for a user to upgrade their system and all of a sudden they now need to set OTEL_SDK_DISABLED=false because we allowed the default to change."*
- *"If there does exist a valid example where a boolean variable needs to have it's default value changed, [...] it still makes sense from the user perspective to keep static defaults. A new variable would be defined and the old variable deprecated."*

Arguments **in favor of adopting Position A for non-environment variable (e.g. declarative configuration) usages** are mentioned on [this comment](https://github.com/open-telemetry/opentelemetry-specification/issues/4344#issuecomment-2548586678) and [this comment](https://github.com/open-telemetry/opentelemetry-specification/issues/4344#issuecomment-2555020477):
- *The boolean env var convention is sensible enough, and following it when introducing new component properties (i.e. not env vars) is the path of least resistance for contributors.*
- *having different conventions for different configuration interfaces (env vars vs. programmatic) is a bad design. I'd like to see us consistent (to the extent it makes sense),*

Arguments **in favour of Position B** include:

- This is inconsistent with other parts of the ecosystem including the Collector, the Java Agent and .NET (see below for usages). Some examples of 'enabled-by-default but not named `disabled` configuration toggles in the Collector are listed [here](https://github.com/open-telemetry/opentelemetry-specification/issues/4344#issuecomment-2550764224).
- *When exposing things as properties double negative seems way counter-intuitive for the end users* (see https://github.com/open-telemetry/opentelemetry-specification/issues/4344#issuecomment-3412628121). This was argued for in #4351.
- This leads to poorer discoverability: users don't know whether a configuration flag is named e.g. `disabled` or `enabled` if they don't also know what the value of it is.
- Position C may not make sense for all usages in the Collector since there may be 'no good default value' for all options or default values may not follow the stabilization framework (e.g. whether people want exemplars on their trace metrics is not a question of stability but rather of resource usage or backend support).

Arguments **in favor of Position C** were mostly discussed on the [2025-10-28 Specification SIG meeting](https://zoom.us/rec/play/OyBhrQo8v5G9wIOoTULFHkUWveIKDUUK25Z7JtKzuNG-UZHvLNj5lDilVP5ECbheR2P-A4a3xA_tGL3V.r2kHFNnsb1f0oaXQ?eagerLoadZvaPages=&isReferralProgramEnabled=false&isReferralProgramAvailable=false&accessLevel=meeting&canPlayFromShare=true&from=share_recording_detail&continueMode=true&componentName=rec-play&originRequestUrl=https%3A%2F%2Fzoom.us%2Frec%2Fshare%2FADgT4LJyaDvkb8q_KytrI6XHMapR95IPpdQGuDVXdWnIu6U6YNYjT1Lx-wdmPl_y.nSsWwRK_MIP5CV1g), discussion starts at around 13:18:
- Mentioned around the 33 minute mark This option has the benefits of Position B but avoids changing the default value of settings, which is one of the arguments in favor of Position A.

----

*Original text of the issue before 2025-11-05.*

The [`MeterConfig`'s `disabled` parameter](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk.md#meterconfig) (introduced in #3877) conflicts with the Collector convention which is to use `enabled` (this is not documented, [this search](https://github.com/search?q=repo%3Aopen-telemetry%2Fopentelemetry-collector+OR+repo%3Aopen-telemetry%2Fopentelemetry-collector-contrib+mapstructure%3A%22enabled%22+language%3AGo&type=code) shows many places where this is used in the YAML configuration). Using `enabled` should be considered for consistency across the whole project.

Another argument against `disabled: false`  is that it has a double negative which is harder to parse. IIRC this is the reason we use `enabled` instead in the Collector configuration.

`disabled: false` however is more idiomatic for Go because of `false` being the zero value of booleans.

cc @jack-berg @codeboten

**Comment by marcalff** [comment]
The fact that boolean environment variables have a default of `false` comes from the spec.

See this conversation:

* https://github.com/open-telemetry/opentelemetry-specification/pull/2755#discussion_r962171923

**Comment by jack-berg** [comment]
We've carried the boolean env var convention into the naming / defaults of component properties. The boolean env var convention is sensible enough, and following it when introducing new component properties (i.e. not env vars) is the path of least resistance for contributors.

Another example of following this guidance where the property doesn't have a standard env variable are [these prometheus exporter properties](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk_exporters/prometheus.md#configuration): `without_units`, `without_type_suffix`, `without_scope_info`, `without_target_info`

> This is not an environment variable though, I am not sure I see the relation

By this logic, the collector YAML config you reference is unrelated the`MeterConfig.disabled`, since YAML config != programmatic config 😉. I think the fact of the matter is that programmatic config, YAML config, and env var config are all intertwined with each other, and conventions that apply to one tend to carry into others.

For example, we carry many of the characteristics of programmatic and env var properties into [declarative config](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/configuration/README.md#declarative-configuration), including naming and defaults.

**Comment by mx-psi** [comment]
> By this logic, the collector YAML config you reference is unrelated theMeterConfig.disabled, since YAML config != programmatic config 😉. I think the fact of the matter is that programmatic config, YAML config, and env var config are all intertwined with each other, and conventions that apply to one tend to carry into others.

I guess the point I didn't understand is that we still intend to add environment variables to configure the SDK even when we have the declarative configuration. If that were not the case, then it could make sense to not consider the "env argument".

I think this only makes sense to reconsider if:
1. This is the first `disabled` example
2. We don't intend to add further environment variables

I suspect (2) is not the case, but could you confirm?

**Comment by pellared** [comment]
> conflicts with the Collector convention which is to use `enabled`

Isn't the case for the collector that all usages of `enabled` are `false` by default so the user has to explicitly enable them? If so then Collector also follows the spec:

> The fact that boolean environment variables have a default of `false` comes from the spec.

**Comment by mx-psi** [comment]
> Isn't the case for the collector that all usages of enabled are false by default so the user has to explicitly enable them? If so then Collector also follows the spec:

No, for example these are all enabled:
- https://github.com/open-telemetry/opentelemetry-collector/blob/4cf491babd7c019f4bd534ce4f5e016073b1d1c1/exporter/exporterqueue/config.go#L30
- https://github.com/open-telemetry/opentelemetry-collector/blob/4cf491babd7c019f4bd534ce4f5e016073b1d1c1/config/configretry/backoff.go#L16
- https://github.com/open-telemetry/opentelemetry-collector/blob/4cf491babd7c019f4bd534ce4f5e016073b1d1c1/exporter/exporterhelper/internal/queue_sender.go#L43
- https://github.com/open-telemetry/opentelemetry-collector/blob/4cf491babd7c019f4bd534ce4f5e016073b1d1c1/exporter/exporterbatcher/config.go#L64

**Comment by pellared** [comment]
> Another argument against `disabled: false` is that it has a double negative which is harder to parse. IIRC this is the reason we use `enabled` instead in the Collector configuration.

For the same reason the OpenTelemetry .NET Automatic Instrumentation has `_ENABLED` env vars even if the default value is `true`. E.g. `OTEL_DOTNET_AUTO_RESOURCE_DETECTOR_ENABLED`. See: https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/blob/main/docs/config.md. We got enough feedback to not go with the  `_DISABLED` env vars.

**Comment by trask** [comment]
my 2c from https://github.com/open-telemetry/opentelemetry-specification/pull/2755#discussion_r966486350:

> I prefer `*enabled` flags over `*disabled` flags (even if that means the default value is true), and we use this pattern for all boolean flags in the otel java agent

**Comment by jack-berg** [comment]
So there are two questions that seem relevant to answer:

1. Do we want to re-litigate the boolean env var naming convention decided in #2755?
2. Do we like following the env var guidance in the programmatic configuration interface?

From my perspective, having different conventions for different configuration interfaces (env vars vs. programmatic) is a bad design. I'd like to see us consistent (to the extent it makes sense), which here seems to mean either accepting the convention, or re-litigating the discussion. On re-ligating: the environment variable spec doc along with the boolean convention is stable. The specific sentence that we're arguing about uses normative "SHOULD". Coupled with the fact that there are very few env vars which follow this advice, I think there's technically some wiggle room to reverse course.

> All Boolean environment variables SHOULD be named and defined such that false is the expected safe default behavior. Renaming or changing the default value MUST NOT happen without a major version upgrade.

With that said, I don't think re-litigating this is a good use of time, so I would personally not participate in that discussion (or block any result).

**Comment by mx-psi** [comment]
> From my perspective, having different conventions for different configuration interfaces (env vars vs. programmatic) is a bad design

We are not just building the specification, we are also building the Collector, the Java Agent, the .NET instrumentation... I think based on your argument, *any* option is a bad design for OpenTelemetry as a whole, since we are bound to have an inconsistency somewhere (be it within the specification or within the wider official OTel ecosystem).

**Comment by jack-berg** [comment]
>  I think based on your argument, any option is a bad design for OpenTelemetry as a whole, since we are bound to have an inconsistency somewhere (be it within the specification or within the wider official OTel ecosystem).

Any inconsistency is bad design, IMO. But the degree of bad depends on the component. I.e. its more problematic if we're not even be able to maintain consistency between configuration interfaces defined within the spec. Ideally, SIGs which are not explicitly bound to the guidance of the specification reach the conclusion that consistency is important.

**Comment by trask** [comment]
just sharing a data point, I suspect we won't change `OTEL_INSTRUMENTATION_[NAME]_ENABLED` in the Java agent, since some of the instrumentations are enabled by default and some are disabled by default, and so it would feel weird (and inconsistent in a different way) to have some specific instrumentations using `OTEL_INSTRUMENTATION_[NAME]_ENABLED` and other specific instrumentations using `OTEL_INSTRUMENTATION_[NAME]_DISABLED`

(and an additional wrinkle is that we have occasionally changed the default for specific instrumentations from enabled to disabled in major version bumps)

**Comment by mx-psi** [comment]
We discussed this on the Specification SIG meeting on 2025-01-07. Some arguments raised (other than the ones already raised on this thread):
- On the 'default should be false' convention for environment variables
  - The number of environment variables that follow the 'default should be false' convention, excluding language-specific conventions is just one: `OTEL_SDK_DISABLED`. We also have one environment variable that does not follow this convention (`OTEL_EXPORTER_OTLP_INSECURE`), that probably precedes the convention
  - We are not obligated to follow the 'default should be false' convention for declarative configuration. There would not be inconsistency between the declarative configuration and the environment variable configuration for this because there will not be a `MeterConfig.disabled` environment variable configuration
  - The spec wording is a 'SHOULD', so we can have exceptions (e.g. for `ENABLED`, to avoid double negation).
  - To some people this convention is less important than being clear on 'enabled'
- We could change any environment variables that use 'DISABLED' to 'ENABLED' by having both and establishing what should happen when both environment variables are defined. This would be backwards compatibility and ensure consistency.
- Every option probably leads to some inconsistency, we need to figure out what is best for end-users. 
- "[Telemetry should be easy](https://opentelemetry.io/community/mission/#telemetry-should-be-easy)" is an argument for prioritizing solving this inconsistency (inconsistency is hard!)

**Comment by mx-psi** [comment]
I filed #4351 which is one of the approaches discussed yesterday. I am not sure if this would need a sponsor or if I should file a separate issue for this change, feel free to take a look at the PR, but I will keep it as draft until someone can help me with clarifying sponsorship/process.

**Comment by trask** [comment]
> I am not sure if this would need a sponsor

this is a relatively small (if not uncontroversial) change, so I don't believe a sponsor is needed

**Comment by MrAlias** [comment]
> We discussed this on the Specification SIG meeting on 2025-01-07. Some arguments raised (other than the ones already raised on this thread):
> 
> * On the 'default should be false' convention for environment variables
>   
>   * The number of environment variables that follow the 'default should be false' convention, excluding language-specific conventions, is just one: `OTEL_SDK_DISABLED`. We also have one environment variable that does not follow this convention (`OTEL_EXPORTER_OTLP_INSECURE`), that probably precedes the convention
>   * We are not obligated to follow the 'default should be false' convention for declarative configuration. There would not be inconsistency between the declarative configuration and the environment variable configuration for this because there will not be a `MeterConfig.disabled` environment variable configuration
>   * The spec wording is a 'SHOULD', so we can have exceptions (e.g. for `ENABLED`, to avoid double negation).
>   * To some people this convention is less important than being clear on 'enabled'
> * We could change any environment variables that use 'DISABLED' to 'ENABLED' by having both and establishing what should happen when both environment variables are defined. This would be backwards compatible and ensure consistency.
> * Every option probably leads to some inconsistency, we need to figure out what is best for end-users.
> * "[Telemetry should be easy](https://opentelemetry.io/community/mission/#telemetry-should-be-easy)" is an argument for prioritizing solving this inconsistency (inconsistency is hard!)

This is a very one-side summary of the meeting conversation. It should be noted for reader here there were dissenting views to many of these points that are not captured. It should not be inferred that consensus was reached.

**Comment by mx-psi** [comment]
> This is a very one-side summary of the meeting conversation. It should be noted for readers here there were dissenting views to many of these points that are not captured. It should not be inferred that consensus was reached.

It was not my intention to claim that there was consensus. As I say on my message I only made a summary of some arguments, *"other than the ones already raised on this thread"*. The main argument against was the re-litigation (which is already mentioned above). If I am missing something else important, feel free to let me know and I can edit my message.

**Comment by lenin-jaganathan** [comment]
Adding comments from https://github.com/open-telemetry/opentelemetry-specification/pull/3877#discussion_r2412477400

> When SDKs/Distros try to expose this as a configuration, 'enabled' seems natural and straightforward. Saying is enabled or not (enable=true seems natural as opposed to saying disabled=false, would say it is true for enabled=false vs disabled=true)

When exposing things as properties double negative seems way counter-intuitive for the end users. I agree with @trask [comment](https://github.com/open-telemetry/opentelemetry-specification/issues/4344#issuecomment-2552712569) that default should not dictate enabled vs disabled naming rather the usability should be considered. 

I see that this issue has lot of discussions around architecturally aligning this decision consistently throughout OpenTelemetry space. Was there any consensus on that?

**Comment by trask** [comment]
In case it helps to have specific examples, here are some of the enabled properties from the OpenTelemetry Java agent which default to true:

### Enabling / disabling the entire Java agent

- `otel.javaagent.enabled`

### Enabling / disabling individual instrumentation

- `otel.instrumentation.common.default-enabled`
- `otel.instrumentation.[name].enabled` *(100+ instrumentations)*
  - with the following exceptions that default to `false`
    - `otel.instrumentation.jdbc-datasource.enabled`
    - `otel.instrumentation.dropwizard-metrics.enabled`
    - `otel.instrumentation.micrometer.enabled`
    - `otel.instrumentation.mybatis.enabled`
    - `otel.instrumentation.spring-boot-actuator-autoconfigure.enabled`
    - `otel.instrumentation.runtime-telemetry-java17.enabled`
    - `otel.instrumentation.oshi.enabled`

### Enabling / disabling individual instrumentation features

- `otel.instrumentation.common.db-statement-sanitizer.enabled`
  - `otel.instrumentation.jdbc.statement-sanitizer.enabled` (defaults to `...db-statement-sanitizer.enabled`)
  - `otel.instrumentation.mongo.statement-sanitizer.enabled` (defaults to `...db-statement-sanitizer.enabled`)
  - `otel.instrumentation.r2dbc.statement-sanitizer.enabled` (defaults to `...db-statement-sanitizer.enabled`)

**Comment by mx-psi** [comment]
I spent a bit of time writing down the usages of enabled and disabled on the Collector components and the OpenTelemetry declarative configuration.

# Collector usage 

## Usages of `enabled`

This may not be an exhaustive list but it should be pretty complete. I did not include options that are named e.g. `tls_enabled`. I do not include the default value but it varies widely, some are enabled by default and some are not:

- Any receiver using mdatagen to generate its config for metric configuration. This includes [55 components](https://github.com/search?q=repo%3Aopen-telemetry%2Fopentelemetry-collector-contrib+%22The+following+metrics+are+emitted+by+default%22+language%3AMarkdown+path%3A%2F%5Ereceiver%5C%2F%2F&type=code)
- Any component using `configretry` (because of the [backoff config](https://github.com/open-telemetry/opentelemetry-collector/blob/ee0bf522ea7c2f60b49332ce3a08dcec4ed8e2c1/config/configretry/backoff.go#L29)). This includes pretty much all exporters, which also have the [queuebatch configuration](https://github.com/open-telemetry/opentelemetry-collector/blob/ee0bf522ea7c2f60b49332ce3a08dcec4ed8e2c1/exporter/exporterhelper/internal/queuebatch/config.go#L20)
- Any component using `configtls` (because of the [TPM module](https://github.com/open-telemetry/opentelemetry-collector/blob/ee0bf522ea7c2f60b49332ce3a08dcec4ed8e2c1/config/configtls/tpm.go#L18)). This includes any component spinning up an HTTP or gRPC server or acting as an HTTP or gRPC client.
- Any component using the [`consumerretry`](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/ea52fa64e3cc07cb57d358762ccf0d146f3461a2/internal/coreinternal/consumerretry/config.go#L13) helper, including the filelog receiver or the syslog receiver
- Any component using the [`resourcetotelemetry` helper](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/ea52fa64e3cc07cb57d358762ccf0d146f3461a2/pkg/resourcetotelemetry/resource_to_telemetry.go#L24) (which includes e.g. the Prometheus exporter and the Datadog exporter)
- [`connectors::exceptions::exemplars::enabled`](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/ea52fa64e3cc07cb57d358762ccf0d146f3461a2/connector/exceptionsconnector/config.go#L21)
- [`connector::spanmetrics::exemplars::enabled`](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/ea52fa64e3cc07cb57d358762ccf0d146f3461a2/connector/spanmetricsconnector/config.go#L110) and [`connector::spanmetrics::events::enabled`](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/ea52fa64e3cc07cb57d358762ccf0d146f3461a2/connector/spanmetricsconnector/config.go#L131)
- [`exporters::coralogix::rate_limiter::enabled`](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/ea52fa64e3cc07cb57d358762ccf0d146f3461a2/exporter/coralogixexporter/config.go#L106)
- [`exporters::datadog::host_metadata::enabled`](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/ea52fa64e3cc07cb57d358762ccf0d146f3461a2/pkg/datadog/config/host.go#L50)
- [`exporters::elasticsearch::logstash_format::enabled`](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/ea52fa64e3cc07cb57d358762ccf0d146f3461a2/exporter/elasticsearchexporter/config.go#L133) as well as three others
- [`exporters::file::group_by::enabled`](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/ea52fa64e3cc07cb57d358762ccf0d146f3461a2/exporter/fileexporter/config.go#L82)
- [`exporters::googlecloudpubsub::ordering::enabled`](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/ea52fa64e3cc07cb57d358762ccf0d146f3461a2/exporter/googlecloudpubsubexporter/config.go#L56)
- [`exporters::prometheusremotewrite::target_info::enabled`](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/ea52fa64e3cc07cb57d358762ccf0d146f3461a2/exporter/prometheusremotewriteexporter/config.go#L67) as well as [`exporters::prometheusremotewrite::remote_write_queue::enabled`](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/ea52fa64e3cc07cb57d358762ccf0d146f3461a2/exporter/prometheusremotewriteexporter/config.go#L77)
- [`exporters::splunkhec::telemetry::enabled`](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/ea52fa64e3cc07cb57d358762ccf0d146f3461a2/exporter/splunkhecexporter/config.go#L58)
- [`exporters::influxdb::v1_compatibility::enabled`](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/ea52fa64e3cc07cb57d358762ccf0d146f3461a2/exporter/influxdbexporter/config.go#L20)
- [`extensions::cgroupruntime::gomaxprocs::enabled`](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/ea52fa64e3cc07cb57d358762ccf0d146f3461a2/extension/cgroupruntimeextension/config.go#L16) as well as [`extensions::cgroupruntime::gomemlimit::enabled`](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/ea52fa64e3cc07cb57d358762ccf0d146f3461a2/extension/cgroupruntimeextension/config.go#L22) 
- [`extensions::healthcheck::check_collector_pipeline::enabled`](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/ea52fa64e3cc07cb57d358762ccf0d146f3461a2/extension/healthcheckextension/config.go#L69)
- [`extensions::zpages::expvar::enabled`](https://github.com/open-telemetry/opentelemetry-collector/blob/ee0bf522ea7c2f60b49332ce3a08dcec4ed8e2c1/extension/zpagesextension/config.go#L26)
- [`processor::redaction::db_sanitizer::<db engine name>::enabled`](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/ea52fa64e3cc07cb57d358762ccf0d146f3461a2/processor/redactionprocessor/internal/db/config.go#L17) 7 in total, plus [1 more](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/ea52fa64e3cc07cb57d358762ccf0d146f3461a2/processor/redactionprocessor/internal/url/config.go#L7)
- [`processors::coralogix::transactions::enabled`](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/ea52fa64e3cc07cb57d358762ccf0d146f3461a2/processor/coralogixprocessor/config.go#L8)
- [`processors::isolationforest::adaptive_window::enabled`](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/ea52fa64e3cc07cb57d358762ccf0d146f3461a2/processor/isolationforestprocessor/config.go#L39)
- [`processors::sumologic::nested_attributes::enabled`](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/ea52fa64e3cc07cb57d358762ccf0d146f3461a2/processor/sumologicprocessor/nesting_processor.go#L17) and [4 more ](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/ea52fa64e3cc07cb57d358762ccf0d146f3461a2/processor/sumologicprocessor/log_fields_conversion_processor.go#L23)
- [`receivers::azuremonitor::dimensions::enabled`](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/ea52fa64e3cc07cb57d358762ccf0d146f3461a2/receiver/azuremonitorreceiver/config.go#L236)
- [`receivers::kafka::metrics::kafka_receiver_records_delay::enabled`](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/ea52fa64e3cc07cb57d358762ccf0d146f3461a2/receiver/kafkareceiver/config.go#L180)
- [`receivers::mongoatlasdb::alerts::enabled`](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/ea52fa64e3cc07cb57d358762ccf0d146f3461a2/receiver/mongodbatlasreceiver/config.go#L42) and [`receivers::mongoatlasdb::logs::enabled`](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/ea52fa64e3cc07cb57d358762ccf0d146f3461a2/receiver/mongodbatlasreceiver/config.go#L56)
- [`receivers::prometheus::api_server::enabled`](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/ea52fa64e3cc07cb57d358762ccf0d146f3461a2/receiver/prometheusreceiver/config.go#L236)
- [`receivers::receivercreator::discovery::enabled`](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/ea52fa64e3cc07cb57d358762ccf0d146f3461a2/receiver/receivercreator/config.go#L93)

## Usages of `disabled`

A single usage, [`exporters::otelarrow::arrow::disabled`](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/ea52fa64e3cc07cb57d358762ccf0d146f3461a2/exporter/otelarrowexporter/config.go#L91)

# OpenTelemetry declarative configuration

## Usages of `enabled`

No usages were found.

## Usages of `disabled`

`disabled `is used in 9 different places, which can all be seen in [the kitchen sink example](https://github.com/open-telemetry/opentelemetry-configuration/blob/main/examples/kitchen-sink.yaml).

**Comment by mx-psi** [comment]
We discussed this again towards the second half of the 2025-10-21 specification meeting.

@MrAlias mentioned (correct me if I misunderstood) that there were user reports of confusion when not following the 'false-by-default' convention. Since from the Collector and Java side we do not seem to have user reports about that confusion, it would be useful to have these user reports to understand the points of confusion. @MrAlias, would you be able to share any public user reports of this confusion?

**Comment by codeboten** [comment]
This issue was discussed in the configuration working group meeting on 27-OCT and the decision there was that this issue not be blocking for the stabilization of the schema. The schema could always add an `enabled` property in the future and implementations could enforce that `enabled` and `disabled` be mutually exclusive when configured.

We already have a precedent in the configuration schema of properties that need to be resolved with things like `headers` and `headers_list`.

In the discussion we also agreed that the configuration schema only has one property (the SDK disabled one) where the use of disabled is significant as the others are under experimental features. We'll happily follow the direction of the spec once a decision has been made.

**Comment by pellared** [comment]
In .NET Automatic Instrumentation we have env vars with `_ENABLED` suffix and there is none `_DISABLED` env var. Most of the `_ENBELED` env vars have a `true` default value. We had a discussion whether we should follow the spec recommendation:

- https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/issues/2090
- https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/pull/2088

If I recall correctly, everyone except from myself preferred to follow the Java and Collector conventions rather than following the spec recommendations.

Having that said **I still prefer the specification recommendation**. I think that having all boolean env vars defaulting to `false` (there were a lot of reasons which I do not want to repeat).

_A spec-compliant alternative could be to use an enumeration configuration instead, e.g. `state` with possible values `enabled` or`disabled` (or `on` and `off` as I like shorter values; `state: off` is even shorter than `disabled: true`). But then people would need to check what are the default values of `state` and it may be an overkill. I also think that `disabled` and `enabled` (defaulting to `false`) are more "natural". PS. I am not happy that I am coming up with another proposal (I imagine it could have been already proposed). Again, I still prefer the specification recommendation._

**Comment by mx-psi** [comment]
I edited the [original comment](https://github.com/open-telemetry/opentelemetry-specification/issues/4344#issue-2744483026) to make a summary of the current positions around this topic, their current usages and the main arguments that I am aware of that have been argued for each item. If I forgot or misrepresented something, let me know and I will amend it. If you want to add something new I am also happy to add it.

I also filed #4723 to explicitly say that the guidance surrounding boolean values for environment variables is not applicable to other configuration interfaces. This is not a solution to this issue but, I feel like, is something that has broad agreement.

I will wait for further work on this apart from that PR until the OTEPs proposed open-telemetry/opentelemetry.io/pull/8208 have been completed, since I think that will inform the discussion about what I have called in the original comment Position C.

**Comment by svrnm** [comment]
blocked by the OTEPs (#4813)

**Comment by svrnm** [comment]
Still blocked by #4813.

**Comment by svrnm** [comment]
#4813 was closed, can this be unblocked? Is this still relevant?

**Comment by jack-berg** [comment]
I think this is resolved, unless I misunderstand the scope:

The [declarative config schema](https://github.com/open-telemetry/opentelemetry-configuration/blob/main/opentelemetry_configuration.json) has boolean properties in a number of places.

The top level `$.disabled` property is the only one that includes disabled.

There are 5 properties that include "enabled". 

There are 4 properties which are boolean and use neither "enabled" or "disabled" in the name.

At this point, there is no convention either implicit or explicit to name boolean properties "disabled".
